### PR TITLE
Add migration step to remove old repo_indexer_status orphaned records

### PR DESCRIPTION
**Pull request by guillep2k**
Migration to remove orphaned records, as corrected in #7734

### CODE REVIEW

**Comment by guillep2k** [comment]
I'd create a backport for v1.9, but I don't know how to resolve the version numbering collision.

**Comment by guillep2k** [comment]
Oh! By the way, I've only tested this with PostgreSQL. I don't know if there's anything particular to take care of with any of the other engines about the "LEFT OUTER" join operation. Perhaps more tests are required?

File: models/migrations/v91.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+import (
+	"code.gitea.io/gitea/models"
+
+	"github.com/go-xorm/xorm"
+)
+
+func removeLingeringIndexStatus(x *xorm.Engine) error {
+
+	var orphaned []*models.RepoIndexerStatus
```
Comment (lafriks) [inline models/migrations/v91.go]: It is better to use type copied into this function otherwise it could start to fail if columns are changed in the future

Comment (zeripath) [reply models/migrations/v91.go]: I.e. copy 

https://github.com/go-gitea/gitea/blob/5b902e2368fdb67eda4ba033f7167fadf72e4350/models/repo_indexer.go#L23-L27

above this function definition.

Comment (guillep2k) [reply models/migrations/v91.go]: Done

File: models/migrations/v91.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+	err := x.
+		Join("LEFT OUTER", "`repository`", "`repository`.id = `repo_indexer_status`.repo_id").
+		Where("`repository`.id is null").
+		Find(&orphaned)
+	if err != nil {
+		return err
+	}
+
+	for _, o := range orphaned {
+		if _, err = x.Delete(o); err != nil {
```
Comment (lunny) [inline models/migrations/v91.go]: It's better to use `x.ID(o.ID).Delete(new(RepoIndexerStatus))`, otherwise it will generate many conditions.

**Review by lunny — CHANGES_REQUESTED** [review]

File: models/migrations/v91.go (hunk 2 above)
Comment (guillep2k) [reply models/migrations/v91.go]: Done

**Review by zeripath — APPROVED** [review]

**Review by zeripath — COMMENTED** [review]
Although you might just be able to do this with a single SQL delete with an appropriate where

**Comment by guillep2k** [comment]
> 
> 
> Although you might just be able to do this with a single SQL delete with an appropriate where

I agree, but I'm not familiar with xorm or its limitations in a multi-engine implementation, and I couldn't find clear examples for the NotIn() operator. Anyway, I figured there shouldn't be too many orphaned records to delete, anyway.

**Comment by lafriks** [comment]
Something like this should do the job:
```go
func removeLingeringIndexStatus(x *xorm.Engine) error {
	return x.Exec(builder.Delete(builder.NotIn("`repo_id`", builder.Select("`id`").From("`repository`"))).From("`repo_indexer_status`"))
}
```

**Comment by lafriks** [comment]
rename migration file to `v92.go`

**Comment by lafriks** [comment]
btw did you test it? I did write it just in comment without testing it ;)

**Comment by guillep2k** [comment]
> 
> 
> btw did you test it? I did write it just in comment without testing it ;)

Yes, of course. 👍

**Comment by lafriks** [comment]
> Yes, of course. 👍

I should start changing code right in production without testing it 😆 I was quite sure I missed some brackets :)

**Comment by guillep2k** [comment]
OK, I've renamed 

> 
> 
> > Yes, of course. 👍
> 
> I should start changing code right in production without testing it 😆 I was quite sure I missed some brackets :)
Only the return type. Anyway, I've tested the functionality as well, not only good compilation.

**Review by lafriks — APPROVED** [review]

**Review by lunny — APPROVED** [review]

**Comment by lafriks** [comment]
LG-TM where are you?

**Comment by lafriks** [comment]
> Only the return type. Anyway, I've tested the functionality as well, not only good compilation.

I did not mean that you did not test I was just surprised for my own comment that I did so close without testing if it even compiles ;)

**Pull request merged**